incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,148 @@
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def get_actions_distribution(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
+        """Get the action predicted by the strategy.
+
+        If past_actions is an int, it is considered a metacoalition id.
+        Otherwise, it is considered as a list of coalitions of the original game.
+        """
+        metacoalition = past_actions if isinstance(past_actions, int) else self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        if sum(self.regret[rank]) == 0:
+            return np.ones(self.number_of_coalitions) / self.number_of_coalitions  # TODO: not all viable!
+        return self.regret[rank] * (self.regret[rank] > 0) / sum(self.regret[rank] * self.regret[rank] > 0)
+
+    def apply_regret(self, bottom_losses: np.ndarray, used_actions: list[list[Coalition]]) -> None:
+        """Scan every possible action sequence, and update the regret of all the regret minimizers.
+
+        Arguments:
+            bottom_losses: the loss of the bottom layer of the game tree.
+                    this should be an array of size (number_of_coalitions,).
+            used_actions: the list of list of actions used to generate the bottom_losses.
+        """
+        used_metacoalition_ranks = [self.meta_id_to_rank[self.get_metacoalition_id(x)]
+                                    for x in used_actions]
+        losses = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                          dtype=RMValue)
+        experienced_losses = np.zeros(self.viable_metacoalitions, dtype=RMValue)
+        experienced_losses[used_metacoalition_ranks] = bottom_losses
+        for i in range(self.number_of_regret_minimizers - 1, -1, -1):
+            metacoalition = self.meta_rank_to_id[i]
+            coalition_pids = list(Coalition(metacoalition).players)
+            next_coalition_pids = [x for x in range(self.number_of_coalitions) if x not in coalition_pids]
+            next_metacoalitions = [Coalition.from_players(coalition_pids + [x]).id for x in next_coalition_pids]
+            next_metacoalition_ranks = self.meta_id_to_rank[next_metacoalitions]
+            losses[i, next_coalition_pids] = experienced_losses[next_metacoalition_ranks]
+            experienced_losses[i] = (losses[i] * self.get_actions_distribution(int(metacoalition))).sum()
```

Comment:
```
metacoalition_strategy = self.get_actions_distribution(int(metacoalition))  # To be called `regret_matching`
experienced_losses[i] = (losses[i] * metacoalition_strategy).sum()
weight = self.iteration if plus else 1  # The average strategy uses linear averaging
self.cumulative_strategy[metacoalition] += weight * metacoalition_strategy
```

Reply:
This seems weird. I don't see the `weight` in the RM+ lecture...

Reply:
The `weight` doesn't have to be there and it is still guaranteed to converge. But this should still have convergence guarantees and it is empirically faster. In fact, RM+ empirically converges in last-iterate, so we might not care about averaging. 

But one thing we should care about is that the cumulative strategy should be multiplied by the reach probability of this node. And that might be annoying to do here...

Reply:
What's "reach probability"? That means like under the current policy, ie the current average strategy? That would be annoying lol

Reply:
Under current policy. The counterfactual value just sums rewards from the bottom of the tree, but the addition to the cumulative strategy (the thing which if you normalize you get the average strategy with convergence guarantees) must be multiplied by the reach probability under current strategy. 

The idea is that the empirical average converges - so if I choose a random _time_ t<T, I should get a strategy in the tree which converges. This is not the same as picking the state. Assume T=2 and budget two. For t=1 you always pick a large coalition at first, and in the state where anything is revealed, you pick something small. Then for t=2, you pick small at first, and large later. Now what is the average prob of choosing small, given you already chose small (i.e. the second budget step). The average there is 50/50, but that is not what the empirical average would give us. Anytime we would actually visit the "I selected small at first" state we picked large.

Reply:
Added.

Reply:
If I understand the code right, you are computing the reach using the average strategy. This is wrong, you should use the current strategy. So when going down the tree, you should just call `self.regret_matching_strategy(int(next_metacoalition))` (I think), and update the reach probability.

Reply:
Fixed.